src/services/GraphQLService.ts

```diff
@@ -1,48 +1,233 @@
+        }
+        return ApiBase.queryGraph(query.replace(/\s\s+/g, ' ').trim(), variables);
+    }
+
+    subscribe(query: string): Observable<object> {
+      const id = this.generateOperationId();
+      this.subscriptions.push(new Subscription(id, query));
+      this.sendMessage(id, MessageType.GQL_START, {query: `subscription { ${query} }`});
+
+      return new Observable<object>(observer => {
+          this.subscriptionHandlerMap[id] = (e: object) => observer.next(e);
+
+          return () => this.stopSubscription(id);
+      });
+    }
+
+    /**
+     * Initialize the EventsService by setting the API key.
+     * When the API key is set, the socket can be opened.
+     * This method is automatically called when doing Qminder.setKey().
+     * @hidden
+     */
+    setKey(apiKey: string) {
+        this.apiKey = apiKey;
+    }
+
+    /**
+     * Set the WebSocket hostname the GraphQL service uses.
+     * @hidden
+     */
+    setServer(apiServer: string) {
+        this.apiServer = apiServer;
+    }
+
+    private stopSubscription(id: string) {
+        console.log(`Stopping subscription ${id}`);
+        this.sendMessage(id, MessageType.GQL_STOP, null);
+
+        delete this.subscriptionHandlerMap[id];
+        this.subscriptions = this.subscriptions.filter((sub) => {
+            return sub.id !== id;
+        });
+    }
+
+    private openSocket() {
+        if (this.connectionStatus != ConnectionStatus.DISCONNECTED) {
+            return;
+        }
+        this.connectionStatus = ConnectionStatus.CONNECTING;
+        const socket = new WebSocket(`${this.apiServer}/graphql/subscription?rest-api-key=${this.apiKey}`);
+        this.socket = socket;
+
+        socket.onopen = () => {
+            console.log('[GraphQL subscription] Connection established!');
+            this.connectionStatus = ConnectionStatus.CONNECTED;
+            this.connectionRetries = 0;
+            this.sendMessage(undefined, MessageType.GQL_CONNECTION_INIT, null);
+        };
+
+        socket.onclose = (event: { code: number }) => {
+            // NOTE: if the event code is 1006, it is any of the errors in the list here:
+            // https://www.w3.org/TR/websockets/#concept-websocket-close-fail
+            console.log('[GraphQL subscription] Connection lost: ' + event.code);
+            this.connectionStatus = ConnectionStatus.DISCONNECTED;
+            this.socket = null;
+
+            // If it wasn't a client-side close socket, retry connecting.
+            if (event.code !== 1000) {
+                // Increase the retry timeout, the more times we retry
+                const timeoutMult = Math.floor(this.connectionRetries / 10);
+                const newTimeout = Math.min(5000 + timeoutMult * 1000, 60000);
+
+                if (this.retryTimeout) {
+                    clearTimeout(this.retryTimeout);
+                }
+
+                console.log('[Qminder Events API] Reconnecting in ' + newTimeout/1000 + ' seconds...');
```

Comment:
These are important enough to be log statements, but need to unify the prefix in square brackets.